### DISCUSSION TITLE
Microservice API gateway

### DISCUSSION THREAD

**Discussion by kachar** [podkrepi-bg/frontend#260] [General]
We'll need an api gateway to distribute the traffic between the microservices and outside world.

It will be best if this gateway can natively handle the most popular transports like `REST`, `GraphQL` and `GRPC` as our service depend on those.

Components and features:

- authentication
- monitoring
- logging
- security
- acl
- caching
- rate-limiting

Candidates: 

### Overview

- https://istio.io/
- https://github.com/istio/community
- https://istio.io/latest/docs/concepts/observability/

![image](https://user-images.githubusercontent.com/893608/121748354-799ac200-cb11-11eb-9a30-6c131fa05eb0.png)

### Overview
- https://konghq.com/install/#kong-community
- https://github.com/Kong/kong
- https://docs.konghq.com/hub/kong-inc/key-auth/
- https://github.com/supabase/supabase/blob/master/docker/dockerfiles/kong/kong.yml

![image](https://user-images.githubusercontent.com/893608/121748182-35a7bd00-cb11-11eb-930a-3b98955cd114.png)

![image](https://user-images.githubusercontent.com/893608/121751886-b7024e00-cb17-11eb-8182-0b7518248ada.png)

**Comment by cefothe** [comment]
Some more open-source API gateways examples https://geekflare.com/api-gateway/#anchor-kong-gateway

**Comment by igoychev** [comment]
I'm starting to be in favor of Kong/Kuma Service mesh, because Istio looks more complicated and less documented than Kuma. Also Istio is Kubernetes only, while Kuma can run on mixed environment with docker and virtual machines.

The github stars speak in favor of Istio though: 27K vs 2.2K for Kuma

Some refererences for Kong/Kuma - these are also supported in Istio too.
-  Kuma ServiceMesh can serve as both internal service mesh and external gateway: https://konghq.com/blog/kuma-service-mesh-gateway-mode/
- supports GRPC: https://kuma.io/docs/1.1.6/policies/protocol-support-in-kuma/ https://konghq.com/blog/manage-grpc-services-kong/
- Observability of services through Prometheus: https://konghq.com/blog/observability-for-your-kubernetes-microservices-using-kuma-and-prometheus/
- Kong Gateway can be integrated with Keykcloak if needed: https://faun.pub/securing-the-application-with-kong-keycloak-101-e25e0ae9ec56
